query_optimizer/rules/UnnestSubqueries.hpp

```diff
@@ -0,0 +1,219 @@
+
+  logical::LogicalPtr apply(const logical::LogicalPtr &input) override {
+    expressions::UnorderedAttributeSet inner_attributes;
+    std::vector<expressions::AttributeReferencePtr> probe_join_attributes;
+    std::vector<expressions::AttributeReferencePtr> build_join_attributes;
+    std::vector<expressions::PredicatePtr> non_hash_join_predicates;
+
+    const logical::LogicalPtr output =
+        applyInternal(input,
+                      &inner_attributes,
+                      &probe_join_attributes,
+                      &build_join_attributes,
+                      &non_hash_join_predicates);
+
+    probe_join_attributes_->insert(probe_join_attributes_->end(),
+                                   probe_join_attributes.begin(),
+                                   probe_join_attributes.end());
+    build_join_attributes_->insert(build_join_attributes_->end(),
+                                   build_join_attributes.begin(),
+                                   build_join_attributes.end());
+    correlated_non_hash_join_predicates_->insert(correlated_non_hash_join_predicates_->end(),
+                                                 non_hash_join_predicates.begin(),
+                                                 non_hash_join_predicates.end());
+
+    return output;
+  }
+
+ private:
+  logical::LogicalPtr applyInternal(const logical::LogicalPtr &input,
+                                    expressions::UnorderedAttributeSet *inner_attributes,
+                                    std::vector<expressions::AttributeReferencePtr> *probe_join_attributes,
+                                    std::vector<expressions::AttributeReferencePtr> *build_join_attributes,
+                                    std::vector<expressions::PredicatePtr> *non_hash_join_predicates);
+
+  logical::LogicalPtr applyToNode(
+      const logical::LogicalPtr &input,
+      expressions::UnorderedAttributeSet *inner_attributes,
+      std::vector<expressions::AttributeReferencePtr> *probe_join_attributes,
+      std::vector<expressions::AttributeReferencePtr> *build_join_attributes,
+      std::vector<expressions::PredicatePtr> *non_hash_join_predicates);
+
+  // For each input expression, call UnnestSubqueriesForExpession to
+  // eliminate any subquery expressions in it and add a join if necessary.
+  logical::LogicalPtr eliminateNestedScalarQueries(const logical::LogicalPtr &node);
+
+  // Eliminate outer attribute references in this node, add predicates with
+  // outer attribute references into join attributes or non_hash_join_predicates.
+  expressions::ExpressionPtr eliminateOuterAttributeReference(
+      const expressions::ExpressionPtr &expression,
+      expressions::UnorderedAttributeSet *inner_attributes,
+      std::vector<expressions::AttributeReferencePtr> *probe_join_attributes,
+      std::vector<expressions::AttributeReferencePtr> *build_join_attributes,
+      std::vector<expressions::PredicatePtr>* non_hash_join_predicates);
```

Comment:
What do you think if the pointer sigil standard applies to here?

`std::vector<expressions::PredicatePtr> *non_hash_join_predicates`.